File: server/services/astrology/rules/ketuConditions.js
/* Ketu-related conditions: traditionally read when Ketu occupies the 1st, 7th or
   9th house, or is conjunct the Moon. */
'use strict';
const { houseFrom } = require('./helpers');

function evaluate(view) {
  const evidence = [], evidenceHi = [];
  let strength = 0;
  const K = view.planets.ketu;
  const h = houseFrom(K, view.lagnaSign);
  if ([1, 7, 9].includes(h)) { evidence.push('Ketu occupies the ' + h + ordinal(h) + ' house from the lagna.'); evidenceHi.push('केतु लग्न से ' + ({ 1: 'प्रथम', 7: 'सप्तम', 9: 'नवम' }[h]) + ' भाव में स्थित है।'); strength += 1; }
  if (view.planets.moon.sign === K.sign) { evidence.push('The Moon is conjunct Ketu, a traditionally strong Ketu influence on the mind.'); evidenceHi.push('चंद्र की केतु से युति है — पारंपरिक रूप से मन पर केतु का प्रबल प्रभाव माना जाता है।'); strength += 1; }

  if (!evidence.length) return null;
  return {
    detected: true,
    severity: strength >= 2 ? 'high' : 'low',
    confidence: Math.min(0.8, 0.5 + 0.15 * strength),
    evidence,
    evidenceHi
  };
}

function ordinal(n) { return n === 1 ? 'st' : n === 2 ? 'nd' : n === 3 ? 'rd' : 'th'; }

module.exports = { evaluate };

File: backend-python/app/services/astrology/rules/ketu_conditions.py
"""Ketu-related conditions — port of rules/ketuConditions.js: traditionally read
when Ketu occupies the 1st, 7th or 9th house, or is conjunct the Moon."""
from .helpers import house_from, ordinal

HOUSE_HI = {1: 'प्रथम', 7: 'सप्तम', 9: 'नवम'}


def evaluate(view: dict, now=None, partner=None) -> dict | None:
    evidence, evidence_hi = [], []
    strength = 0
    K = view['planets']['ketu']
    h = house_from(K, view['lagnaSign'])
    if h in (1, 7, 9):
        evidence.append('Ketu occupies the ' + str(h) + ordinal(h) + ' house from the lagna.')
        evidence_hi.append('केतु लग्न से ' + HOUSE_HI[h] + ' भाव में स्थित है।')
        strength += 1
    if view['planets']['moon']['sign'] == K['sign']:
        evidence.append('The Moon is conjunct Ketu, a traditionally strong Ketu influence on the mind.')
        evidence_hi.append('चंद्र की केतु से युति है — पारंपरिक रूप से मन पर केतु का प्रबल प्रभाव माना जाता है।')
        strength += 1

    if not evidence:
        return None
    return {
        'detected': True,
        'severity': 'high' if strength >= 2 else 'low',
        'confidence': min(0.8, 0.5 + 0.15 * strength),
        'evidence': evidence,
        'evidenceHi': evidence_hi,
    }
